source/extensions/filters/network/zookeeper_proxy/decoder.cc

```diff
@@ -476,41 +476,101 @@ void DecoderImpl::skipStrings(Buffer::Instance& data, uint64_t& offset) {
 }
 
 Network::FilterStatus DecoderImpl::onData(Buffer::Instance& data) {
-  return decodeAfterBuffer(data, DecodeType::READ);
+  return decodeAndBuffer(data, zk_filter_read_buffer_, DecodeType::READ);
 }
 
 Network::FilterStatus DecoderImpl::onWrite(Buffer::Instance& data) {
-  return decodeAfterBuffer(data, DecodeType::WRITE);
+  return decodeAndBuffer(data, zk_filter_write_buffer_, DecodeType::WRITE);
 }
 
-Network::FilterStatus DecoderImpl::decodeAfterBuffer(Buffer::Instance& data, DecodeType dtype) {
+Network::FilterStatus DecoderImpl::decodeAndBuffer(Buffer::Instance& data,
+                                                   Buffer::OwnedImpl& zk_filter_buffer,
+                                                   DecodeType dtype) {
+  uint32_t data_len = data.length();
+  uint32_t zk_filter_buffer_len = zk_filter_buffer.length();
+
+  if (zk_filter_buffer_len > 0) {
+    if (zk_filter_buffer_len + data_len > INT_LENGTH) {
+      // Prepend ZooKeeper filter buffer to current buffer.
+      data.prepend(zk_filter_buffer);
+      decodeAndBufferHelper(data, zk_filter_buffer, dtype);
+      // Drain the prepended ZooKeeper filter buffer.
+      data.drain(zk_filter_buffer_len);
+    } else {
+      zk_filter_buffer.add(data);
+    }
+  } else if (zk_filter_buffer_len == 0) {
+    decodeAndBufferHelper(data, zk_filter_buffer, dtype);
+  }
+  return Network::FilterStatus::Continue;
+}
+
+void DecoderImpl::decodeAndBufferHelper(Buffer::Instance& data, Buffer::OwnedImpl& zk_filter_buffer,
+                                        DecodeType dtype) {
+  uint32_t data_len = data.length();
+  uint32_t zk_filter_buffer_len = zk_filter_buffer.length();
   uint64_t offset = 0;
-  int32_t len = 0;
+  uint32_t len = 0;
+  // Boolean to check whether there is at least one full packet in the buffer.
+  bool has_full_packets = false;
 
-  while (offset < data.length()) {
+  while (offset < data_len) {
     try {
       // Peek packet length.
       len = helper_.peekInt32(data, offset);
-      ensureMinLength(len, INT_LENGTH + XID_LENGTH);
+      if (dtype == DecodeType::READ) {
```

Comment:
i prefer `ensureMinLength(len, dtype == DecodeType::READ ? XID_LENGTH + INT_LENGTH : XID_LENGTH + ZXID_LENGTH + INT_LENGTH);`

you could ASSERT(dtype == READ || dtype == WRITE) at the beginning if you wan tto prevent more dtype